### PR TITLE
Kiruththigan

### PR DESCRIPTION
**Pull request by kiruththigan**

### CODE REVIEW

File: defect-tracker-server/src/main/java/com/defect/tracker/controller/DefectController.java [hunk 1]
```diff
@@ -93,4 +95,13 @@
 				validationFailureStatusCodes.getDefectById()), HttpStatus.BAD_REQUEST);
 	}
 
+	@GetMapping(value = "/count/{projectName}")
```
Comment (ramz7) [inline defect-tracker-server/src/main/java/com/defect/tracker/controller/DefectController.java]: add the value in EndPoint class

File: defect-tracker-server/src/main/java/com/defect/tracker/services/DefectServiceImpl.java [hunk 2]
```diff
@@ -49,4 +54,20 @@ public boolean isDefectExistsById(Long id) {
 		return defectRepository.existsById(id);
 	}
 
+	@Override
+	public DefectCountByPriotryAndProjectDto countByProject(String projectName) {
+		Project project = projectService.findByName(projectName);
+		DefectCountByPriotryAndProjectDto defectCountDto = new DefectCountByPriotryAndProjectDto();
+		defectCountDto.setPriorityHigh(defectRepository.countByPriorityAndProject("high", project));
```
Comment (ramz7) [inline defect-tracker-server/src/main/java/com/defect/tracker/services/DefectServiceImpl.java]: can use the project name directly

File: defect-tracker-server/src/main/java/com/defect/tracker/controller/DefectController.java [hunk 3]
```diff
@@ -35,14 +37,16 @@
...
+	private EmailService emailservice;
+
 	@PostMapping(value = EndpointURI.DEFECT)
-	public ResponseEntity<Object> addDefect(@Valid @RequestBody DefectDto defectDto) {
-		if (defectService.isDefectExistsById(defectDto.getId())) {
-			return new ResponseEntity<>(new ValidationFailureResponse(ValidationConstance.DEFECT_EXISTS,
-					validationFailureStatusCodes.getDefectExistsById()), HttpStatus.BAD_REQUEST);
-		}
+	public ResponseEntity<Object> addDefect(@Valid @RequestBody DefectDto defectDto) throws MessagingException {
+		System.out.println(defectDto.getAssignedToId());
```
Comment (ramz7) [inline defect-tracker-server/src/main/java/com/defect/tracker/controller/DefectController.java]: delete system.out.println

File: defect-tracker-server/src/main/java/com/defect/tracker/controller/DefectController.java [hunk 4]
```diff
@@ -35,14 +37,16 @@
...
 		Defect defect = mapper.map(defectDto, Defect.class);
 		defectService.createDefect(defect);
+		System.out.println(defect.getAssignedTo() + "not work");
```
Comment (ramz7) [inline defect-tracker-server/src/main/java/com/defect/tracker/controller/DefectController.java]: delete this print

File: defect-tracker-server/src/main/java/com/defect/tracker/data/dto/DesignationDto.java [hunk 5]
```diff
@@ -7,14 +7,13 @@
 
 public class DesignationDto {
 	private Long id;
-	
-	@NotNull(message="name is not null")
-	@NotBlank(message="name is not blank")
-	@NotEmpty(message="name is not empty")
-	@Pattern(regexp = "^[a-zA-Z.\\-\\/+=@_ ]*$",message="name not allowed special character and number")
+
+	@NotNull(message = "name is not null")
```
Comment (ramz7) [inline defect-tracker-server/src/main/java/com/defect/tracker/data/dto/DesignationDto.java:11]: use validationMessage.properties file as i used in projectDto

Comment (ramz7) [reply defect-tracker-server/src/main/java/com/defect/tracker/data/dto/DesignationDto.java:11]: @Thayanthi20

File: defect-tracker-server/src/main/java/com/defect/tracker/data/dto/TypeDto.java [hunk 6]
```diff
@@ -8,11 +8,11 @@
 public class TypeDto {
 
 	private Long id;
-	
-	@NotNull(message="name is not null")
-	@NotBlank(message="name is not blank")
-	@NotEmpty(message="name is not empty")
-	@Pattern(regexp = "^[a-zA-Z.\\-\\/+=@_ ]*$",message="name not allowed special character and number")
+
+	@NotNull(message = "name is not null")
```
Comment (ramz7) [inline defect-tracker-server/src/main/java/com/defect/tracker/data/dto/TypeDto.java:12]: @mathi-14 
@Mithila0907

File: defect-tracker-server/src/main/java/com/defect/tracker/services/DefectServiceImpl.java (hunk 2 above)
Comment (ramz7) [reply defect-tracker-server/src/main/java/com/defect/tracker/services/DefectServiceImpl.java]: you can use countByPriorityAndProjectName("high",projectName)

File: defect-tracker-server/src/main/java/com/defect/tracker/data/entities/CustomUserDetails.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	}
+
+	@Override
+	public boolean isCredentialsNonExpired() {
+		return true;
+	}
+
+	@Override
+	public boolean isEnabled() {
+		// return super.isEnabled();
```
Comment (ramz7) [inline defect-tracker-server/src/main/java/com/defect/tracker/data/entities/CustomUserDetails.java]: delete this

File: defect-tracker-server/src/main/java/com/defect/tracker/data/repositories/EmployeeRepository.java [hunk 8]
```diff
@@ -10,5 +12,7 @@
 
 	boolean existsByEmail(String email);
 
-	Employee findByEmail(String email);
+	//Employee findByEmail(String email);
```
Comment (ramz7) [inline defect-tracker-server/src/main/java/com/defect/tracker/data/repositories/EmployeeRepository.java]: delete this

File: defect-tracker-server/src/main/java/com/defect/tracker/services/CustomUserDetailsService.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+	@Autowired
+	private EmployeeRepository employeeRepository;
+
+	@Override
+	public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+		Optional<Employee> employeeOpt = employeeRepository.findByEmail(username);
+		employeeOpt
+		.orElseThrow(() -> new UsernameNotFoundException("username not found"));
+		
+		/*UserDetails user=Employee.withUserName(employeeOpt.get().getEmail())
```
Comment (ramz7) [inline defect-tracker-server/src/main/java/com/defect/tracker/services/CustomUserDetailsService.java]: delete this

File: defect-tracker-server/src/main/java/com/defect/tracker/util/EndpointURI.java [hunk 10]
```diff
@@ -36,14 +36,9 @@
...
-	// Designation
-	public static final String DESIGNATION =BASE_API_PATH +"Designation" ;
-	public static final String DESIGNATION_BY_ID = BASE_API_PATH + "Designation" +ID;
-
-	// Project Allocation
-	public static final String PROJECT_ALLOCATION = BASE_API_PATH + "projectAllocation";
-	public static final String PROJECT_ALLOCATION_BY_ID = BASE_API_PATH + "projectAllocation" + ID;
-	public static final String MODULE = BASE_API_PATH + "module";
+	// Type
+	public static final String TYPE = BASE_API_PATH + "Type";
```
Comment (ramz7) [inline defect-tracker-server/src/main/java/com/defect/tracker/util/EndpointURI.java:43]: don't use capital T

File: defect-tracker-server/src/main/java/com/defect/tracker/config/JwtAuthenticationEntryPoint.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.security.web.AuthenticationEntryPoint;
+import org.springframework.stereotype.Component;
+
+@Component
+public class JwtAuthenticationEntryPoint implements AuthenticationEntryPoint {
+
+	@Override
+	public void commence(HttpServletRequest request, HttpServletResponse response,
+			AuthenticationException authException) throws IOException {
+
```
Comment (ramz7) [inline defect-tracker-server/src/main/java/com/defect/tracker/config/JwtAuthenticationEntryPoint.java]: remove line break

File: defect-tracker-server/src/main/java/com/defect/tracker/config/JwtRequestFilter.java [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+		} else {
+			logger.warn("JWT Token does not begin with Bearer String");
+		}
+
+		if (username != null && SecurityContextHolder.getContext().getAuthentication() == null) {
+
+			UserDetails userDetails = this.customUserDetailsService.loadUserByUsername(username);
+
+			if (jwtTokenUtil.validateToken(jwtToken, userDetails)) {
+
```
Comment (ramz7) [inline defect-tracker-server/src/main/java/com/defect/tracker/config/JwtRequestFilter.java]: remove line break

File: defect-tracker-server/src/main/resources/ValidationMessages.properties [hunk 13]
```diff
@@ -48,6 +48,7 @@ employee.designationId.blank=807050
 validation.module.AlreadyExists=000000
 validation.project.AlreadyExists=100000
 validation.projectName.AlreadyExists=1000003
+validation.project.NotAvalable=1000004
```
Comment (ramz7) [inline defect-tracker-server/src/main/resources/ValidationMessages.properties:53]: this is for? is it NotExists?

File: defect-tracker-server/src/main/java/com/defect/tracker/controller/DefectController.java [hunk 14]
```diff
@@ -95,7 +95,7 @@
 				validationFailureStatusCodes.getDefectById()), HttpStatus.BAD_REQUEST);
 	}
 
-	@GetMapping(value = "/count/{projectName}")
+	@GetMapping(value = EndpointURI.DEFECT_COUNT_BY_PROJECT_NAME)
```
Comment (ramz7) [inline defect-tracker-server/src/main/java/com/defect/tracker/controller/DefectController.java]: Priority_severity_defect_count_by_project_name

File: defect-tracker-server/src/main/java/com/defect/tracker/services/DefectServiceImpl.java (hunk 2 above)
Comment (kiruththigan) [reply defect-tracker-server/src/main/java/com/defect/tracker/services/DefectServiceImpl.java]: face the error beacause getter and setter type is Project

File: defect-tracker-server/src/main/java/com/defect/tracker/util/EndpointURI.java (hunk 10 above)
Comment (kiruththigan) [reply defect-tracker-server/src/main/java/com/defect/tracker/util/EndpointURI.java:43]: @Thayanthi20 
@Mithila0907

File: defect-tracker-server/src/main/resources/ValidationMessages.properties (hunk 13 above)
Comment (kiruththigan) [reply defect-tracker-server/src/main/resources/ValidationMessages.properties:53]: @Banulipar

File: defect-tracker-server/src/main/java/com/defect/tracker/services/DefectServiceImpl.java (hunk 2 above)
Comment (ramz7) [reply defect-tracker-server/src/main/java/com/defect/tracker/services/DefectServiceImpl.java]: did you implement like this? defectCountDto.setPriorityHigh(defectRepository.countByPriorityAndProjectName("high", projectName));

**Pull request merged**